File: dynamic_programming/list_and_array/2_num_sum.java
package dynamic_programming.list_and_array;


class ListNode{
    int val;
    ListNode next;
    ListNode(){}
    ListNode(int val){
        this.val = val;
    }
    ListNode(int val, ListNode next){
        this.val = val;
        this.next = next;
    }
}
class Solution{
    public ListNode addTowNumber(ListNode l1, ListNode l2){
        ListNode head = null, tail = null;
        // tail指向当前链表的尾部
        int carry = 0;
        while(l1 != null ||l2 != null){
            // 若l1或l2不为空，则计算当前位相加的结果
            int n = l1 != null ? l1.val : 0;//这里注意的是Java判断必须是boolean类型，不能用l1直接作为判断条件
            // 在c++中，如果l1为空
            // int n = l1 ? l1.val : 0;
            // 就可以这样简写，但是java不行
            int m = l2 != null ? l2.val : 0;
            int sum = n + m + carry;
            // 下面将结果放入新节点中
            if(head == null){
                head = tail = new ListNode(sum%10);
            }else{
                tail.next = new ListNode(sum%10);
                tail = tail.next;
            }
            carry = sum/10;
            if(l1 != null){
                l1 = l1.next;
            }
            if(l2 != null){
                l2 = l2.next;
            }
        }
        if(carry == 1){
            tail.next = new ListNode(1);
        }
        return head;
    }
}

File: dynamic_programming/list_and_array/2-num-sum.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode*head = nullptr,*tail = nullptr;
        int carry = 0;
        while(l1||l2){
            int n1 = l1?l1->val:0;//若l1存在，则将l1中存储的值放入n1中
            int n2 = l2?l2->val:0;
            int sum = n1+n2+carry;
            if(!head){//若头部为空，将创建第一个节点
                head = tail = new ListNode(sum%10);
            }
            else{
                tail->next = new ListNode(sum%10);
                tail = tail->next;
            }
            carry = sum /10;//记录下进位的值
            if(l1){
                l1 = l1 ->next;//将进行下一位的运算
            }
            if(l2){
                l2 = l2->next;
            }
            
        }
        if(carry>0){//这里是考虑最后一位的进位
                tail->next = new ListNode(carry);
        }
        return head;
    }
};